File: python/03-copy-paste-payroll/src/payroll_calculator.py
class PayrollCalculator:
    def __init__(self):
        pass

    def generate_payslips(self, employees):
        payslips = []
        for emp in employees:
            gross = deductions = net = 0
            if emp["type"] == "fulltime":
                gross = emp["salary"] / 12
                deductions = gross * 0.25
                if emp.get("bonus"):
                    gross += emp["bonus"] / 12
                net = gross - deductions
            elif emp["type"] == "parttime":
                gross = emp["hours"] * emp["rate"]
                deductions = gross * 0.15
                net = gross - deductions
            elif emp["type"] == "contract":
                gross = emp["flatFee"]
                deductions = gross * 0.1
                net = gross - deductions
            payslips.append({"id": emp["id"], "name": emp["name"], "type": emp["type"], "gross": round(gross, 2), "deductions": round(deductions, 2), "net": round(net, 2)})
        return payslips

File: ruby/03-copy-paste-payroll/src/payroll_calculator.rb
# frozen_string_literal: true

class PayrollCalculator
  def generate_payslips(employees)
    payslips = []
    employees.each do |emp|
      gross = deductions = net = 0
      case emp[:type]
      when 'fulltime'
        gross = emp[:salary] / 12.0
        deductions = gross * 0.25
        gross += emp[:bonus] / 12.0 if emp[:bonus]
        net = gross - deductions
      when 'parttime'
        gross = emp[:hours] * emp[:rate]
        deductions = gross * 0.15
        net = gross - deductions
      when 'contract'
        gross = emp[:flat_fee]
        deductions = gross * 0.1
        net = gross - deductions
      end
      payslips << { id: emp[:id], name: emp[:name], type: emp[:type], gross: gross.round(2),
                    deductions: deductions.round(2), net: net.round(2) }
    end
    payslips
  end
end
